### PR TITLE
7112 make back off limit configurable

### PR DESCRIPTION
**Pull request by sheidkamp**
# Description

Expose backoffLimit, completions, manualSelector, and parallelism fields for k8s jobs

# Context

Users needed this feature to set these fields' values in the install process

# Checklist:

- [X] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/7112

### CODE REVIEW

File: install/test/helm_test.go [hunk 1]
```diff
@@ -5684,6 +5696,33 @@ metadata:
...
 
+				// Positive/default values of setTtlAfterFinished are tested in the regular test
+				DescribeTable("setTtlAfterFinished=false suppresses ttlSecondsAfterFinished on Jobs",
+					func(kind string, resourceName string, jobValuesPrefix string, extraArgs ...string) {
+						prepareMakefile(namespace, helmValues{
+							valuesArgs: append([]string{
+								jobValuesPrefix + ".setTtlAfterFinished=false",
+								jobValuesPrefix + ".ttlSecondsAfterFinished=42",
+							}, extraArgs...),
+						})
```
Comment (jackstine) [inline install/test/helm_test.go:5707]: I think we should have a test if `setTtlAfterFinished` is set to `true`.

File: install/helm/gloo/templates/6.5-gateway-certgen-job.yaml [hunk 2]
```diff
@@ -18,9 +18,7 @@ metadata:
     "helm.sh/hook-weight": "10"
     "helm.sh/hook-delete-policy": hook-succeeded
 spec:
-  {{- with .Values.gateway.certGenJob.activeDeadlineSeconds }}
-  activeDeadlineSeconds: {{ . }}
-  {{- end }}
+  {{- include "gloo.jobSpecStandardFields" .Values.gateway.certGenJob | nindent 2 -}}
```
Comment (jackstine) [inline install/helm/gloo/templates/6.5-gateway-certgen-job.yaml:21]: I think we have to get rid of line 66.
```
  {{- if .Values.gateway.certGenJob.setTtlAfterFinished }}
  ttlSecondsAfterFinished: {{ .Values.gateway.certGenJob.ttlSecondsAfterFinished }}
  {{- end }}
```

File: install/helm/gloo/templates/5-resource-cleanup-job.yaml [hunk 3]
```diff
@@ -16,9 +16,7 @@ metadata:
     "helm.sh/hook-weight": "5" # run this job after the role/rolebinding is created
     "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
 spec:
-  {{- with .Values.gateway.cleanupJob.activeDeadlineSeconds }}
-  activeDeadlineSeconds: {{ . }}
-  {{- end }}
+  {{- include "gloo.jobSpecStandardFields" .Values.gateway.cleanupJob | nindent 2 -}}
```
Comment (jackstine) [inline install/helm/gloo/templates/5-resource-cleanup-job.yaml:19]: not sure how this conflicts with line 75
`  ttlSecondsAfterFinished: {{ .Values.gateway.rolloutJob.ttlSecondsAfterFinished }}`

File: install/helm/gloo/templates/_helpers.tpl [hunk 4]
```diff
@@ -65,6 +65,38 @@ priorityClassName: {{ . }}
...
+{{- /* include ttlSecondsAfterFinished if setTtlAfterFinished is undefined or equal to true.
+      The 'kindIs' comparision is how we can check for undefined */ -}}
+{{- if or (kindIs "invalid" .setTtlAfterFinished) .setTtlAfterFinished -}}
+{{- with .ttlSecondsAfterFinished  -}}
+ttlSecondsAfterFinished: {{ . }}
+{{ end -}}
+{{- end -}}
+{{- with .extraPodLabels -}}
+extraPodLabels: {{ . }}
+{{ end -}}
```
Comment (jackstine) [inline install/helm/gloo/templates/_helpers.tpl:94]: aren't these already included in the file `install/helm/gloo/templates/6.5-gateway-certgen-job.yaml` at line 27 - 31.

File: docs/content/reference/values.txt [hunk 5]
```diff
@@ -293,6 +293,10 @@
 |gateway.certGenJob.hostAliases[]|interface|||
 |gateway.certGenJob.initContainers[]|interface||[InitContainers](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#containers) to be added to the array of initContainers on the deployment.|
 |gateway.certGenJob.activeDeadlineSeconds|int||Deadline in seconds for Kubernetes jobs.|
+|gateway.certGenJob.backoffLimit|int||Specifies the number of retries before marking this job failed. Defaults to 6|
```
Comment (sam-heilbron) [inline docs/content/reference/values.txt:296]: Is the default to 6 a kubernetes default, or one that we define somewhere? I couldn't find how we are applying that default

File: install/test/helm_test.go [hunk 6]
```diff
@@ -5652,12 +5652,16 @@ metadata:
 					Entry("resource cleanup job", "Job", "gloo-resource-cleanup", "gateway.cleanupJob"),
 				)
 
-				DescribeTable("can set activeDeadlineSeconds and ttlSecondsAfterFinished on Jobs",
+				DescribeTable("can set JobSpec fields on Jobs",
```
Comment (sam-heilbron) [inline install/test/helm_test.go:5655]: Love these tests!

**Review by sam-heilbron — COMMENTED** [review]
Looking good! One question about default values

File: install/helm/gloo/templates/_helpers.tpl (hunk 4 above)
Comment (sheidkamp) [reply install/helm/gloo/templates/_helpers.tpl:94]: They are, and they require extra processing, so will remove them (and extraPodAnnotations)

File: install/helm/gloo/templates/5-resource-cleanup-job.yaml (hunk 3 above)
Comment (sheidkamp) [reply install/helm/gloo/templates/5-resource-cleanup-job.yaml:19]: It does, removing

File: install/test/helm_test.go (hunk 1 above)
Comment (sheidkamp) [reply install/test/helm_test.go:5707]: Will do

File: docs/content/reference/values.txt (hunk 5 above)
Comment (sheidkamp) [reply docs/content/reference/values.txt]: Its a kubes default, making it clear.

File: install/test/helm_test.go (hunk 6 above)
Comment (sheidkamp) [reply install/test/helm_test.go:5664]: 👍

**Review by sam-heilbron — DISMISSED** [review]

File: install/helm/gloo/generate/values.go [hunk 7]
```diff
@@ -84,6 +84,10 @@ type PodSpec struct {
 type JobSpec struct {
 	*PodSpec
 	ActiveDeadlineSeconds   *int              `json:"activeDeadlineSeconds,omitempty" desc:"Deadline in seconds for Kubernetes jobs."`
+	BackoffLimit            *int              `json:"backoffLimit,omitempty" desc:"Specifies the number of retries before marking this job failed. In kubernetes, defaults to 6"`
+	Completions             *int              `json:"completions,omitempty" desc:"Specifies the desired number of successfully finished pods the job should be run with."`
+	ManualSelector          *bool             `json:"manualSelector,omitempty" desc:"manualSelector controls generation of pod labels and pod selectors."`
```
Comment (jenshu) [inline install/helm/gloo/generate/values.go:89]:
```suggestion
	ManualSelector          *bool             `json:"manualSelector,omitempty" desc:"Controls generation of pod labels and pod selectors."`
```
nit: for consistency with the other descriptions

File: install/test/helm_test.go [hunk 8]
```diff
@@ -5684,6 +5705,58 @@ metadata:
...
+							if u.GetKind() == kind && u.GetName() == resourceName {
+								a := getFieldFromUnstructured(u, append(prefixPath, "spec", "ttlSecondsAfterFinished")...)
+								Expect(a).To(BeNil())
+								return true
+							}
+							return false
+						})
+						Expect(resources.NumResources()).To(Equal(1))
+					},
+					Entry("mtls certgen job", "Job", "gloo-mtls-certgen", "gateway.certGenJob", "global.glooMtls.enabled=true"),
```
Comment (jenshu) [inline install/test/helm_test.go:5730]: can you also add an entry for the `gateway-certgen` job?

File: install/test/helm_test.go [hunk 9]
```diff
@@ -5684,6 +5705,58 @@ metadata:
...
+							if u.GetKind() == kind && u.GetName() == resourceName {
+								a := getFieldFromUnstructured(u, append(prefixPath, "spec", "ttlSecondsAfterFinished")...)
+								Expect(a).To(Equal(int64(TTL_SECONDS_AFTER_FINISHED)))
+								return true
+							}
+							return false
+						})
+						Expect(resources.NumResources()).To(Equal(1))
+					},
+					Entry("mtls certgen job", "Job", "gloo-mtls-certgen", "gateway.certGenJob", "global.glooMtls.enabled=true"),
```
Comment (jenshu) [inline install/test/helm_test.go:5756]: can yo also add an entry for the `gateway-certgen` job?

File: install/helm/gloo/templates/5-resource-migration-job.yaml [hunk 10]
```diff
@@ -16,9 +16,7 @@ metadata:
     "helm.sh/hook-weight": "3" # run this job after the role/rolebinding is created, and before the validation webhook is upgraded
     "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
 spec:
-  {{- with .Values.gateway.rolloutJob.activeDeadlineSeconds }}
-  activeDeadlineSeconds: {{ . }}
-  {{- end }}
+  {{- include "gloo.jobSpecStandardFields" .Values.gateway.rolloutJob | nindent 2 -}}
```
[the remark this answers is not in the document]
Comment (sheidkamp) [reply install/helm/gloo/templates/5-resource-migration-job.yaml:19]: removed

File: install/test/helm_test.go (hunk 8 above)
Comment (sheidkamp) [reply install/test/helm_test.go:5731]: yup

File: install/test/helm_test.go (hunk 9 above)
Comment (sheidkamp) [reply install/test/helm_test.go:5758]: yup

Comment (sheidkamp) [reply install/test/helm_test.go:5758]: yup

**Review by jenshu — APPROVED** [review]
looks great!

**Review by sam-heilbron — APPROVED** [review]

**Pull request merged**